src/backend/distributed/planner/insert_select_planner.c

```diff
@@ -304,6 +309,156 @@ CreateDistributedInsertSelectPlan(Query *originalQuery,
 }
 
 
+/*
+ * CreateInsertSelectIntoLocalTablePlan creates the plan for INSERT .. SELECT queries
+ * where the selected table is distributed and the inserted table is not.
+ *
+ * To create the plan, this function first creates a distributed plan for the SELECT
+ * part. Then puts it as a subquery to the original (non-distributed) INSERT query as
+ * a subquery. Finally, it puts this INSERT query, which now has a distributed SELECT
+ * subquery, in the masterQuery.
+ *
+ * If the SELECT query is a router query, whose distributed plan does not have a
+ * masterQuery, this function also creates a dummy masterQuery for that.
+ */
+DistributedPlan *
+CreateInsertSelectIntoLocalTablePlan(uint64 planId, Query *originalQuery, ParamListInfo
+									 boundParams, bool hasUnresolvedParams,
+									 PlannerRestrictionContext *plannerRestrictionContext)
+{
+	RangeTblEntry *selectRte = ExtractSelectRangeTableEntry(originalQuery);
+
+	Query *selectQuery = BuildSelectForInsertSelect(originalQuery);
+	originalQuery->cteList = NIL;
+	DistributedPlan *distPlan = CreateDistributedPlan(planId, selectQuery, selectQuery,
+													  boundParams, hasUnresolvedParams,
+													  plannerRestrictionContext);
+	if (distPlan->masterQuery == NULL)
+	{
+		distPlan->masterQuery = CreateMasterQueryForRouterPlan(
+			distPlan);
+	}
+
+	/* masterQuery of a distributed select is for combining the results from
+	 * worker nodes on the coordinator node. Putting it as a subquery to the
+	 * insert query makes insert query insert the combined select value from
```

Comment:
These 3 "insert" in quick succession made this sentence a bit confusing to read for me.

```suggestion
	 * INSERT query causes the INSERT query to insert the combined select value from
```